### PR TITLE
[chore] use our own logging implementation

### PR DESCRIPTION
**Pull request by NyaaaWhatsUpDoc**
- replace logrus with our own custom logger with a _very similar_ API but improved performance:
  - the entire log generation process is memory pooled
  - log fields are key-value structs instead of instantiating a hashmap for every single set of fields (a hashmap is not an insignificant data structure!)
- saves having to make odd little workarounds of logrus' library for our required usage, e.g. splitting outputs to stdout/stderr
- allows easier moving to another logging library in the future, we can just update the internals of the functions we're using to instead wrap another log library
- the logging of value types in key-value fields is handled by my own format library which produces much more easily debuggable output (it derefs ptrs instead of just printing e.g. `(*string)(0xDEADBEEF)` ) while keeping performance at a very similar level to that of the standard library format printing (if not faster at times). though this is entirely optional and relying on the standard library instead is as easy as removing the `kvformat` build tag
- still keeps trace-level database logged queries un-escaped for easier copy-pasting during development
- includes calling function with every single log entry, without spamming the logs with extra file/line information. makes it easier to see both flow of logic, and in the case of debugging where an error may be coming from. this is executed in a performant manner by caching function PCs to friendly-formatted function names. The bulk of caller lookup logic comes from the determining of function information, rather than the PC itself (_AFAIK_ ...)

Most of the important changes are in `internal/log`. Other changes around the codebase.

It's also worth noting for those that aren't @tsmethurst (who already knows this), that this is a _proposed_ change. It was something I worked on over the weekend after having a little spark of inspiration and I've been running it on my personal instance for >24hrs at this point.

### CODE REVIEW

**Review by tsmethurst — COMMENTED** [review]
This is cool! I like it :)

A few comments and things that I think you missed while copy-pasting, so those need to be checked and fix, and just one suggestion about changing one of the functions to make the syntax a bit less finnicky

File: internal/db/bundb/account.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ import (
 	"strings"
 	"time"
 
-	"github.com/sirupsen/logrus"
+	log "github.com/dsoprea/go-logging"
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/account.go]: done

File: internal/db/bundb/mention.go [hunk 2]
```diff
@@ -22,7 +22,7 @@ import (
 	"context"
 
 	"codeberg.org/gruf/go-cache/v2"
-	"github.com/sirupsen/logrus"
+	log "github.com/dsoprea/go-logging"
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/mention.go]: done

File: internal/db/bundb/notification.go [hunk 3]
```diff
@@ -22,7 +22,7 @@ import (
 	"context"
 
 	"codeberg.org/gruf/go-cache/v2"
-	"github.com/sirupsen/logrus"
+	log "github.com/dsoprea/go-logging"
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/notification.go]: again

File: internal/db/bundb/status.go [hunk 4]
```diff
@@ -24,7 +24,7 @@ import (
 	"database/sql"
 	"time"
 
-	"github.com/sirupsen/logrus"
+	log "github.com/dsoprea/go-logging"
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/status.go]: pong

File: internal/db/bundb/timeline.go [hunk 5]
```diff
@@ -21,7 +21,7 @@ package bundb
 import (
 	"context"
 
-	"github.com/sirupsen/logrus"
+	log "github.com/dsoprea/go-logging"
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/timeline.go]: pong

File: internal/federation/federatingdb/followers.go [hunk 6]
```diff
@@ -17,12 +18,10 @@ import (
...
 // The library makes this call only after acquiring a lock first.
 func (f *federatingDB) Followers(ctx context.Context, actorIRI *url.URL) (followers vocab.ActivityStreamsCollection, err error) {
-	l := logrus.WithFields(
-		logrus.Fields{
-			"func": "Followers",
-			"id":   actorIRI,
-		},
-	)
+	l := log.WithFields(kv.Fields{
+
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/federation/federatingdb/followers.go:22]: uk

File: internal/log/caller.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+import (
+	"runtime"
+	"strings"
+	"sync"
+)
+
+var (
+	// fnCache is a cache of PCs to their calculated function names.
+	fnCache = map[uintptr]string{}
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/log/caller.go:11]: PC is the program counter :)

File: internal/transport/controller.go [hunk 8]
```diff
@@ -24,12 +24,12 @@ import (
 	"crypto/x509"
 	"encoding/json"
 	"fmt"
+	"log"
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/transport/controller.go]: sorted

File: internal/log/log.go [hunk 9]
```diff
@@ -1,113 +1,259 @@
...
+		sysout.Crit(msg)
 	}
 }
 
-// OutputSplitFunc implements the io.Writer interface for use with Logrus, and simply
-// splits logs between stdout and stderr depending on their severity.
-type OutputSplitFunc func(lvl []byte) io.Writer
+// args returns an args format string of format '%v' * count.
+func args(count int) string {
+	const args = `%v%v%v%v%v%v%v%v%v%v` +
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/log/log.go:257]: it saves requiring separate print and printf implementations :innocent:

File: internal/log/caller.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+			return fn
+		}
+
+		// Fetch frame info for caller pc
+		frame, _ := runtime.CallersFrames(rpc[:]).Next()
+
+		if frame.PC != 0 {
+			name := frame.Function
+
+			// Drop all but the package name and function name, no mod path
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/log/caller.go:63]: it really does improve the readability immensely :p

File: internal/config/validate.go [hunk 11]
```diff
@@ -55,7 +54,7 @@ func Validate() error {
 		// no problem
 		break
 	case "http":
-		logrus.Warnf("%s was set to 'http'; this should *only* be used for debugging and tests!", ProtocolFlag())
+		//log.Warnf("%s was set to 'http'; this should *only* be used for debugging and tests!", ProtocolFlag())
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/config/validate.go]: This was commented out due to an import cycle, though I moved the config.Get...() code out of the log package now which fixes this :)

File: internal/federation/federatingdb/create.go [hunk 12]
```diff
@@ -46,13 +48,9 @@ import (
 // Under certain conditions and network activities, Create may be called
 // multiple times for the same ActivityStreams object.
 func (f *federatingDB) Create(ctx context.Context, asType vocab.Type) error {
-	l := logrus.WithFields(
-		logrus.Fields{
-			"func": "Create",
-		},
-	)
+	l := log.WithFields(kv.Fields{}...)
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/federation/federatingdb/create.go]: It was not :')

**Comment by NyaaaWhatsUpDoc** [comment]
@tsmethurst the only remaining change is the `{K: ..., V: ...}` formatting of fields which i'm about to address

**Comment by tsmethurst** [comment]
This looks good now! I'll have a final look through it tomorrow and see if anything jumps out, then i'll merge it if you're ready :)

**Pull request merged**